### PR TITLE
Firearms Overhaul V 1.0 - Ranges converted from Real Values to Relative in game values

### PR DESCRIPTION
**Pull request by kasanryukin**
# Abstract
Current ballistic calculations in game are not conducive to any firearm beyond a range of 14 tiles with any expectation of accuracy with the exception of the RM11B Sniper Rifle (reason unkown at the moment as it's the only gun with accurate fire at low skill beyond 16 tiles).  Calculations only take two factors into consideration for range: recoil and dispersion.  While both of these are needed numbers in the calculation, effective firing ranges of guns should also be a factor in the calculation before recoil and dispersion are factored in.

Link to Real Data Spreadsheet: https://docs.google.com/spreadsheets/d/1aGH8N2kITxZ_8i1Vyi1G-FIVfQOTL6euvKxZkYOc-FE/edit?usp=sharing
# Outline

The goal of the project is an overhaul in firearms to something both viable in game, but not overtly broken like screwing around with dispersion values.  Part one of the project is assigning the RANGE value to every firearm as defined in item.cpp.  Numbers are based off the values listed near the end of ranged.cpp as Medium (9 tiles) and Short (4 tiles).  Most SMGs and Shotguns fall between the two. Assault rifles, some carbines and some battle rifles, are upgraded to a value of Long (+5 over medium), DMRs/Hunting Rifles/Some Carbines/Some Sniper Rifles, are Upgraded to +10 over medium. Some designated sniper rifles are at +15. Launchers and anti-material weapons were largely left untouched as they already had existing values that exceed the scheme defined above.

Currently, the values don't affect actual accuracy at the extended ranges. This will be altered once code changes are committed that add range values to the math equations for calculating accuracy.

This is very much still a WIP as I'm familiarizing myself with the code related to projectile weaponry.

Edit: Secondary aim of project: Adjust dates of leadworks firearms to match game design.

# Current Project Stage

Ranges have now been converted from Real Ranges to Relative Ranges. Most pistols will behave similar to each other. Shotguns have some variation on range, SMGs run the range between a pistol and home made rifle. Rifles were converted based around 14 tiles being a central range and 500m being the median range to achieve 14 tiles. Depending on the effective range of the real weapon, rifles range between 10 and 20 tiles. DMRs, Sniper Rifles, and some Lever action Rifles, fall in the 18-25 tile range.  RivTech and Anti-material Sniper Systems fall at the range cap of 30 tiles.  Laser weapons had ranges adjusted to fit these values. Some guns had dispersion increased or decreased (no more than 20 MoA) based on RL weapon descriptions.

Ranges adjustments are ready for review before being merged.  Once the data is part of the branch, Kevin can use the ranges as part of any future changes to the firearms systems.

### CODE REVIEW

**Comment by kevingranade** [comment]
As we discussed over DM, we need the data from your sources, otherwise this
isn't a step forward from the numbers being derived via rectal extraction.

**Comment by kasanryukin** [comment]
Data will be available shortly, just copying it over to a google sheet now.

**Comment by BorkBorkGoesTheCode** [comment]
Should I close #22741 ?

**Comment by kasanryukin** [comment]
Maybe merge the two? Let's see what Kevin thinks of the data collected. He can compare it against both PRs

**Comment by kasanryukin** [comment]
For public review of data:
https://docs.google.com/spreadsheets/d/1aGH8N2kITxZ_8i1Vyi1G-FIVfQOTL6euvKxZkYOc-FE/edit?usp=sharing

**Comment by BorkBorkGoesTheCode** [comment]
IRL the .700 Nitro Express seems to have shorter range than the .308 Winchester.

**Comment by Firestorm01X2** [comment]
About distances in code comments. Watch out. It may be eternitybears old and based on old game balance.

**Comment by kasanryukin** [comment]
It's part of the on hit math in the firearms code.

**Comment by Firestorm01X2** [comment]
>Numbers are based off the values listed near the end of ranged.cpp as Medium (9 tiles) and Short (4 tiles).

Also do we realy need this? I mean - lets just say: one tile is 20 meters for ranged wepons and do conversion based on it.
This way we wil lbe talking about real ranges scaled to game. But not: "lets say that 9 tiles- is meadium range".

**Comment by kasanryukin** [comment]
I was informed this wasn't a real possibility due to there being several different systems in place to determine distance or space. 1 tile simply can't be defined as 1 square meter, or even something larger. When you start in on 1 tile being 15 meters, suddenly cars and houses look massively large by comparison.  Something discussed in the DM between Kevin and I was just a homogeneous  category of range based on firearm type, which would be the simplest method of implementation. Thinking in terms of tiles being a certain range instead of specific meters allows the system to loosely let guns operate at ranges typical of their design.

**Comment by Firestorm01X2** [comment]
Ok.
But did you see this: 
https://github.com/CleverRaven/Cataclysm-DDA/issues/21244#issuecomment-310013403

It is more or less current range balance.

**Comment by kasanryukin** [comment]
Because the ranged value was already defined in the code. Also that was Coolthulu's fork to fix/revert some changes.  #21468 is the relevant code in this case, which Kevin is working on.  I've presented the requested data, lets see what Kevin decides to do with it.

**Comment by Firestorm01X2** [comment]
You know. There is one thing. Your pull request will be closed. Chance is around 95%.

**Comment by Zireael07** [comment]
Altering dates in descriptions has nothing to do with the stated goal of the PR. And given that the PR is already large, it makes it even less likely to get merged.

**Comment by Firestorm01X2** [comment]
Is not that dates connected to LORE?

**Comment by Rivet-the-Zombie** [comment]
>Is not that dates connected to game backstory and date?

It is indeed. I'm opposed to those changes in particular.

**Comment by kasanryukin** [comment]
You guys should pay attention on the forums where the date range of the game was actively discussd and our Lord and Savior @kevingranade  said the dates are erroneous in the first place because the game's time frame is today, not 30 years from today.  Edits are in line with his two recent PR's ALSO removing the dates off of items in-game to be more in-line with the design document.

**Comment by Cyrano7** [comment]
I thought it was Whales who said it was near future? Don't quote me on that though. I'd have to go back and re-listen to that rougelike radio episode from ages ago and see if he said anything about that. I can't listen to it now cause i'm at work but here it is. http://www.roguelikeradio.com/2012/01/episode-20-cataclysm.html#more

**Comment by Cyrano7** [comment]
I listened to the beginning and the original creator of the game said it takes place basically 50 years in the future. 

@kasanryukin Do you have a link to that discussion?

**Comment by kasanryukin** [comment]
https://discourse.cataclysmdda.org/t/list-of-ideas-mostly-just-dreams-but-cool/14766/7?u=druidniam

**Comment by Cyrano7** [comment]
I can understand wanting to remove hard dates from descriptions of things. But why replace them with more hard dates? And why 2015? It looks like the intent with the original cataclysm was to have a rolling timeline of about 50 years in the future from 'now'.

Maybe we should remove mentions of specific dates? That's got to be better then replacing them with specific current(ish) day ones.

**Comment by kasanryukin** [comment]
Seems reasonable, but again, as Kevin has stated, the game takes place "Today" Any hard date of "Before today" would be an acceptable date given that statement.

I'll remove the dates, and supplement them with "Early 21st century" to fit with the changes to the laser guns.

**Comment by Cyrano7** [comment]
@kevingranade was the game setting changed to present day at some point? According to whales in 2012 he mention several times it was near-future on rougelike radio. It looks like this message was repeated on the wikipedia page and that rock-paper-shotgun article. I not trying to be confrontational or anything I just want to know when/if/how the switch happened and what's the root of the confusion.

**Comment by kasanryukin** [comment]
If you look at the discourse link I posed above in the comments you can see him state in the thread the setting is "Today".

**Comment by kasanryukin** [comment]
You even asked for the link.

**Comment by Firestorm01X2** [comment]
@kevingranade 
>Numbers are based off the values listed near the end of ranged.cpp as Medium (9 tiles) and Short (4 tiles). Most SMGs and Shotguns fall between the two.

Are this ranges from comments in ranged.cpp still valid?

**Comment by Firestorm01X2** [comment]
Did you take in account fact that most of Leadworks guns has built in scopes. It is in their descritpion.
It should affect dispersion of the weapon.

File: data/json/items/gun/20x66mm.json [hunk 1]
```diff
@@ -46,6 +46,7 @@
     "color": "dark_gray",
     "ammo": "20x66mm",
     "skill": "shotgun",
+    "ramge": 8,
```
Comment (Reclusive-reptile) [inline data/json/items/gun/20x66mm.json]: Typo. Should be "range"

**Comment by kasanryukin** [comment]
Good catch.

**Comment by kasanryukin** [comment]
@Firestorm01X2  "Did you take in account fact that most of Leadworks guns has built in scopes and attachments? It is in their description."

The dispersion on those weapons already reflect their integral equipment.  I only touched dispersion in two places and left all other numbers alone per Kevin's request that nobody screw around with dispersion as a mechanism to accuratize firearms for greater distances.  The hope is that Kevin will come up with a method where dispersion doesn't take effect until perhaps say, 50% of the weapon range is reached, then dispersion starts to be part of the calculation.  All guns get a little more accurate, but only Launchers, AMRs and SRs really get a bit cheesy with range.  Given the rarity of those weapons, I wouldn't see that change as bad.  Ranges can always be tweaked down to fit the framework of the game better rather than as a representation of a real guns range.

**Comment by kasanryukin** [comment]
@kevingranade  Dispersion as a whole might need adjustments then. The "effective" range of weapons used on-point hits (meaning successful shot in a 3'x3' zone) to determine the range with a couple noted exceptions.  In the example of say the SCAR-L, that means hitting a 3 inch target consistently and accurately at 600 meters.

For clarification on what that means: Point shooting is the act of pointing your rifle at a target and depressing the trigger, before you even put an eyeball to sights or options. Just shoulder your weapon and fire. Using Pythagoras and the above gun, a shot taken should at most miss a target at effective range by 7.5" to either side of that target.  I have to assume the shots taken were by a semi-competent shooter since most of these are manufacture specs, and I would assume the person doing the shooting is familiar with the weapon and shooting in general. (Call him a 3/3 with base perception).

As it stands, that gun using the same stats in game at a range of even just 14 tiles, has the chance to miss by as many as 4 tiles to either side. 

Dispersion is a great mechanic as it accurately represents the chance to hit or miss a target, but the miss part of the equations are extremely out of touch with the reality of what the firearms claims are.  I'm not actually sure where your original dispersion values came from, and that might be the source of the issue.

Here's an except from SOCOMs requirements on the above weapon system (paragraph also contained information on the SCAR-H and actually includes hit expectations which I didn't originally find in my research and might be of more useful data.)

The SCAR-L would have to achieve a 70 percent hit ratio on 500 meter point target and a 600 meter area target (T) and 600 meter point target and an 800 meter area target (O). The SCAR-H would have to achieve a 70 percent hit ratio on 600 meter point target and an 800 meter area target (T) and 800 meter point target and a 1000 meter area target (O). The SCAR-L and SCAR-H could not add more than 1 minute of angle (MOA) at 300 meters (T) and not more than .25 MOA at 300 meters (O) to the weapon's shot group.

Evening doubling that, you'd still be .5 MOA too high as the Json defines for the SCAR-L off from max effective range, but on point for an Area Hit.

Note: I think the only reason I could find data like that on the SCAR-L was because it was a contract military project farmed out to various weapons manufactures, and FN Herstal was the one to succeed in meeting those requirements.

This also provides another good point of data. The US Military was looking for 70% accuracy at it's effective range.  If we assume all weapon data I collected also assumed a 70% success at effective range, that gives another useful point of data to work from.

I think I rambled in this comment.

**Comment by kevingranade** [comment]
> @kevingranade Dispersion as a whole might need adjustments then. 

Yes, it does.  
I never claimed the dispersion numbers in the game are in any way accurate, they're mostly arbitrary values chosen to make guns kind of act as expected within the context if the game.  Look at comments like https://github.com/CleverRaven/Cataclysm-DDA/issues/21244#issuecomment-310013403 for the kind of rationale used.  If we can derive more reality-based numbers, I'd love to incorporate them, but they simply weren't available at the time these values were set.  
Again, #22753 is where that is happening.

> In the example of say the SCAR-L, that means hitting a 3 inch target consistently and accurately at 500 meters.
> Point shooting is the act of pointing your rifle at a target and depressing the trigger, before you even put an eyeball to sights or options.

I'm assuming you meant feet here, because 3" at 500 meters is absurd.

What you're saying here is that a shot off the shoulder suffers from a *maximum* of ~36 MoA of dispersion due to being unaimed for an ideal rifle.  That's a very interesting data point.  For comparison the current value used in the game for this scenario is 600 MoA.

> I would assume the person doing the shooting is familiar with the weapon and shooting in general. (Call him a 3/3 with base perception).

That's a very odd assumption to make, I assume both the firearm manufacturer and SOCOM are going to want to use a shooter whose abilities match either a typical or lowest common denominator special forces member.  In either case that's not "familiar with shooting", it's more like "professional shooter", placing skills in the 6-8 area and with above average stats (low perception or dexterity is a DQ for special forces).

I feel like I'm repeating myself here, but I cannot stress this enough.  We CANNOT make guns in-game as accurate as they are irl.  That 36MoA dispersion from not aiming is approximately 2" at the approximately 60 meters that represents maximum engagement range in-game.  Similarly, the majority of guns will have sub-inch deflection at 60 meters, meaning there is no differentiation between them.  In short, that would break the game, so it's not going to happen.

What *can* happen is we set reality-based values in json (the dispersion value for guns and ammo), and we set a compromise between that value and what the game can tolerate in the game engine, and as the game engine is improved to increase engagement range, that compromise is adjusted to take advantage of the additional range.

**Comment by kasanryukin** [comment]
Should I mark this as ready or close it?

**Comment by BorkBorkGoesTheCode** [comment]
@kasanryukin the git is saying that your PR has merge conflicts because it can't cope with both your 308 changes and my 308 changes here: 299105f5d4e5a9b5288f6b16ce8d93345e618c99 This is fixable.

**Comment by kasanryukin** [comment]
@BorkBorkGoesTheCode  Fixed

**Comment by BorkBorkGoesTheCode** [comment]
Do you have a "fix merge conflicts" button n your UI at the bottom of the PR?

**Comment by kasanryukin** [comment]
In theory I fixed it this time.

File: data/json/items/gun/308.json [hunk 2]
```diff
@@ -237,7 +237,7 @@
     "ammo": "308",
     "ranged_damage": -3,
     "magazine_well": 1,
-    "relative": { "range": -3 }
+    "relative": { "range": -3 },
     "magazine_well": 1
   }
```
Comment (BorkBorkGoesTheCode) [inline data/json/items/gun/308.json]: This magazine_well is causing the buildbot to complain about duplicate entries.

**Comment by kasanryukin** [comment]
Ok, looks like that's actually fixed it

**Comment by Firestorm01X2** [comment]
And conflicts again.

**Comment by BorkBorkGoesTheCode** [comment]
I really dislike GitHub because of this stuff.

**Comment by Firestorm01X2** [comment]
After fixing in should be reviewed. Looks like it affects too much json files and definitely will get conflicts again.

**Comment by Night-Pryanik** [comment]
> I really dislike GitHub because of this stuff.

That's not GitHub stuff at all. That's Git stuff.

**Comment by kasanryukin** [comment]
Conflicting files contain original date descriptions that were changed per Kevin's statement on when the game takes place, so I'm not going to edit them to be non-conflicting with the master branch.  Merging won't directly impact firearms in any way until the code itself references any of the changes I made, but it will fix the descriptions to have non-hard set dates on firearm introduction.  Marking this as ready to merge.

**Comment by ZhilkinSerg** [comment]
> Conflicting files contain original date descriptions that were changed per Kevin's statement on when the game takes place, so I'm not going to edit them to be non-conflicting with the master branch. Merging won't directly impact firearms in any way until the code itself references any of the changes I made, but it will fix the descriptions to have non-hard set dates on firearm introduction. Marking this as ready to merge.

Merging won't be possible until all conflicts are resolved. Conflicting cpp/h will prevent successful compilation, conflicting json will cause errors during loading of data during game start.

**Comment by kevingranade** [comment]
The spreadsheet is here for reference, we never had any intention of applying these adjustments this way, so I'm closing to avoid further confusion.

**Pull request closed without merging**